src/bin/ion/commands/beta/generate/generator.rs

```diff
@@ -243,8 +278,8 @@ impl<'a, L: Language> CodeGenerator<'a, L> {
             }
             IslTypeRef::Anonymous(type_def, _) => {
                 self.anonymous_type_counter += 1;
-                self.generate_abstract_data_type(modules, type_def)?;
                 let name = format!("AnonymousType{}", self.anonymous_type_counter);
+                self.generate_abstract_data_type(modules, type_def)?;
```

Comment:
Yes, it does. This method `generate_abstract_data_type` generates the code for given anonymous type, previously it was getting generated before the `anonymous_type_counter` was used for `name` field where `anonymous_type_counter` could change in the generation process of given anonymous type.

Reply:
That's going to be easy for someone to mess up later. Can we refactor this so that it doesn't rely on side-effects? Maybe `generate_abstract_data_type` could accept `name` so that it doesn't have to interact with the type counter?

Reply:
So `generate_abstract_data_type` can be called recursively if there is nested type definition or a type definition has anonymous type. I can surely change it to use a name but it would still interact with type counter whenever it gets a new anonymous type and tries to generate code for it.

Reply:
Right, but the counter can be incremented by a function such as `get_next_anonymous_type_name()` which is responsible for incrementing the counter _and_ returning the name. Then, the only time you ever interact with the counter is specifically when you are requesting a new name, and then you pass around that name to anywhere that needs it instead of having multiple bits of code that read the counter to create it.

Basically, if the ordering of the statements is important because of some shared piece of data, then that requirement should be expressed in the signatures/contracts of the relevant bits of code.